pymc_bart/tree.py

```diff
@@ -52,77 +106,102 @@ class Tree:
             output=np.zeros((num_observations, shape)).astype(config.floatX).squeeze(),
         )
 
-    def __getitem__(self, index):
+    def __getitem__(self, index) -> Node:
         return self.get_node(index)
 
-    def __setitem__(self, index, node):
+    def __setitem__(self, index, node) -> None:
         self.set_node(index, node)
 
-    def copy(self):
-        tree = {
+    def copy(self) -> "Tree":
+        tree: Dict[int, Node] = {
             k: Node(v.index, v.value, v.idx_data_points, v.idx_split_variable)
             for k, v in self.tree_structure.items()
         }
-        return Tree(tree, self.idx_leaf_nodes.copy(), self.output.copy())
+        idx_leaf_nodes = self.idx_leaf_nodes.copy() if self.idx_leaf_nodes is not None else None
+        output = self.output.copy() if self.output is not None else None
+        return Tree(tree_structure=tree, idx_leaf_nodes=idx_leaf_nodes, output=output)
 
-    def get_node(self, index) -> "Node":
+    def get_node(self, index: int) -> Node:
         return self.tree_structure[index]
 
-    def set_node(self, index, node):
+    def set_node(self, index: int, node: Node) -> None:
         self.tree_structure[index] = node
-        if node.is_leaf_node():
-            self.idx_leaf_nodes.append(index)
+        if node.is_leaf_node() and self.idx_leaf_nodes is not None:
+            # self.idx_leaf_nodes.append(index)
+            self.idx_leaf_nodes = np.append(self.idx_leaf_nodes, index)
 
-    def grow_leaf_node(self, current_node, selected_predictor, split_value, index_leaf_node):
+    def grow_leaf_node(
+        self, current_node: Node, selected_predictor: int, split_value: float, index_leaf_node: int
+    ) -> None:
         current_node.value = split_value
         current_node.idx_split_variable = selected_predictor
         current_node.idx_data_points = None
-        self.idx_leaf_nodes.remove(index_leaf_node)
+        if self.idx_leaf_nodes is not None:
+            # self.idx_leaf_nodes.remove(index_leaf_node)
+            self.idx_leaf_nodes = np.setdiff1d(self.idx_leaf_nodes, index_leaf_node)
 
-    def trim(self):
-        tree = {
+    def trim(self) -> "Tree":
+        tree: Dict[int, Node] = {
             k: Node(v.index, v.value, None, v.idx_split_variable)
             for k, v in self.tree_structure.items()
         }
         return Tree(tree, None, None)
 
-    def get_split_variables(self):
+    def get_split_variables(self) -> Generator[int, None, None]:
         for node in self.tree_structure.values():
             if node.is_split_node():
                 yield node.idx_split_variable
 
-    def _predict(self):
+    def _predict(self) -> Optional[npt.NDArray[np.float_]]:
         output = self.output
-        for node_index in self.idx_leaf_nodes:
-            leaf_node = self.get_node(node_index)
-            output[leaf_node.idx_data_points] = leaf_node.value
-        return output.T
 
-    def predict(self, x, excluded=None):
+        if output is None:
```

Comment:
again, isn't self.output always an array?

Reply:
Not always, there is an edge case `Tree(tree, None, None)` https://github.com/pymc-devs/pymc-bart/blob/main/pymc_bart/tree.py#L102  If I am not missing something 🙈

Reply:
(resolved) above.

Reply:
see https://github.com/pymc-devs/pymc-bart/pull/78/commits/8d4fc232cd415b0aa9384d6c5bd506fd2c94518e